src/integrations/dataframes.jl

```diff
@@ -1,155 +1,9 @@
-        else
-            push!(constructor_call.args, :(columns[$i][i]))
-        end
-    end
-
-    quote
-        i = state
-        columns = iter.columns
-        a = $constructor_call
-        return a, state+1
-    end
-end
-
-function Base.done{T,TS}(iter::DataFrameIterator{T,TS}, state)
-    return state>size(iter.df,1)
-end
-
-# Sink
-
-@generated function _filldf(columns, enumerable)
-    n = length(columns.types)
-    push_exprs = Expr(:block)
-    for i in 1:n
-        if columns.parameters[i] <: DataArray
-            ex = :( push!(columns[$i], isnull(i[$i]) ? DataArrays.NA : get(i[$i])) )
-        else
-            ex = :( push!(columns[$i], i[$i]) )
-        end
-        push!(push_exprs.args, ex)
-    end
-
-    quote
-        for i in enumerable
-            $push_exprs
-        end
-    end
-end
-
-function _DataFrame(x)
-    iter = getiterator(x)
-
-    T = eltype(iter)
-    if !(T<:NamedTuple)
-        error("Can only collect a NamedTuple iterator into a DataFrame")
-    end
-
-    column_types = TableTraits.column_types(iter)
-    column_names = TableTraits.column_names(iter)
-
-    columns = []
-    for t in column_types
-        if isa(t, TypeVar)
-            push!(columns, Array{Any}(0))
-        elseif t <: DataValue
-            push!(columns, DataArray(t.parameters[1],0))
-        else
-            push!(columns, Array{t}(0))
-        end
-    end
-    df = DataFrames.DataFrame(columns, fieldnames(T))
-    _filldf((df.columns...), iter)
-    return df
-end
-
-DataFrames.DataFrame{T<:NamedTuple}(x::Array{T,1}) = _DataFrame(x)
-
-function DataFrames.DataFrame(x)
-    if isiterabletable(x)
-        return _DataFrame(x)
-    else
-        return convert(DataFrames.DataFrame, x)
-    end
-end
-
-function DataFrames.ModelFrame(f::DataFrames.Formula, source; kwargs...)
-    isiterabletable(source) || error()
-    return DataFrames.ModelFrame(f, DataFrames.DataFrame(source); kwargs...)
-end
-
-function StatsBase.fit{T<:StatsBase.StatisticalModel}(::Type{T}, f::DataFrames.Formula, source, args...; contrasts::Dict = Dict(), kwargs...)
-    isiterabletable(source) || error()
-    mf = DataFrames.ModelFrame(f, source, contrasts=contrasts)
-    mm = DataFrames.ModelMatrix(mf)
-    y = model_response(mf)
-    DataFrames.DataFrameStatisticalModel(fit(T, mm.m, y, args...; kwargs...), mf, mm)
-end
-
-function StatsBase.fit{T<:StatsBase.RegressionModel}(::Type{T}, f::DataFrames.Formula, source, args...; contrasts::Dict = Dict(), kwargs...)
-    isiterabletable(source) || error()
-    mf = DataFrames.ModelFrame(f, source, contrasts=contrasts)
-    mm = DataFrames.ModelMatrix(mf)
-    y = model_response(mf)
-    DataFrames.DataFrameRegressionModel(fit(T, mm.m, y, args...; kwargs...), mf, mm)
+if isdefined(DataFrames, :Nulls)
+    include("dataframes-null.jl")
+else
+    include("dataframes-dataarray.jl")
 end
```

Comment:
@nalimilan & @quinnj Do you think this idea with ``isdefined...`` will work here? Essentially I will need a version of IterableTables.jl that works with both the old DataArrays.jl based DataFrames.jl and the new Nulls.jl based version at the same time, because I can't sort things out via REQUIRE files given that I used Requires.jl. Other than that, this PR should make Query.jl, IterableTables.jl and the whole ecosystem work with the new DataFrames.jl.

Reply:
I guess that could work (but I don't know anything about Requires.jl). But better test for `isdefined(DataFrames, :null)` since `Nulls` will go away once `null` is defined in Base.